ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt

```diff
@@ -0,0 +1,188 @@
+    override fun getPort() = (channel.remoteAddress as InetSocketAddress?)?.port ?: 0
+    override fun getLocalPort() = (channel.localAddress as InetSocketAddress?)?.port ?: 0
+
+    override fun shutdownInput() {
+        channel.shutdownInput()
+    }
+
+    override fun shutdownOutput() {
+        channel.shutdownOutput()
+    }
+
+    override fun create(stream: Boolean) {
+        require(stream){"only stream sockets are supported"}
+        if (!::channel.isInitialized) {
+            channel = AsynchronousSocketChannel.open()
+        }
+    }
+
+    override fun connect(host: String, port: Int) =
+        connect(InetSocketAddress(host, port), 0)
+
+    override fun connect(address: InetAddress, port: Int) = connect(
+        InetSocketAddress(address, port), 0)
+
+    override fun connect(address: SocketAddress, timeout: Int) {
+        val cf = channel.connect(address)
+        if (timeout > 0) {
+            cf.get(timeout.toLong(), TimeUnit.MILLISECONDS)
+        } else {
+            cf.get()
+        }
+    }
+
+    public
+    override fun getInputStream(): InputStream {
+        return object : InputStream(){
+            override fun read(): Int {
+                val oneByte = ByteArray(1)
+                val rc = read(oneByte, 0, 1)
+                if (rc < 0)
+                    return -1
+                return oneByte[0].toInt()
+            }
+
+            override fun read(b: ByteArray, off: Int, len: Int): Int {
+                synchronized(input) {
+                    inputLock.acquire() // protects async read operation
+                    val count = min(len, input.remaining())
+                    if (count > 0) {
+                        input.get(b, off, count)
+                        inputLock.release()
+                        return count
+                    }
+
+                    input.compact()
+                    val rf = CompletableFuture<Int>()
+                    val to = getOption(SocketOptions.SO_TIMEOUT) as Number
+                    channel.read(input, 0, TimeUnit.MILLISECONDS, rf,
+                        object : CompletionHandler<Int, CompletableFuture<Int>> {
+
+                            override fun completed(result: Int, f: CompletableFuture<Int>) {
+                                if (!f.isDone) {
+                                    f.complete(result)
+                                }
+                                input.flip()
+                                inputLock.release()
+                            }
+
+                            override fun failed(exc: Throwable, f: CompletableFuture<Int>) {
+                                if (!f.isDone) {
+                                    f.completeExceptionally(exc)
+                                }
+                                inputLock.release()
+                            }
+                        })
+
+                    try {
+                        val read = rf.get(to.toLong(), TimeUnit.MILLISECONDS)
```

Comment:
same floor comment wrt timeouts. i expect it's accounted for somewhere but just pointing it out

Reply:
same as above Socket.setSoTimeout checks the value